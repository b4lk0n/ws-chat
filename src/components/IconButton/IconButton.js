import React from 'react';
import { node, oneOf, oneOfType, string } from 'prop-types';
import cx from 'classnames';
import { Icon } from '../Icon';

import styles from './IconButton.module.css';

export const IconButton = ({ icon, size, className, ...props }) => (
  <button {...props} className={cx(styles.root, styles[size], className)}>
    <Icon value={icon} size={size === 'small' ? size : undefined} />
  </button>
);
IconButton.propTypes = {
  icon: oneOfType([string, node]).isRequired,
  size: oneOf(['small', 'large']),
  className: string,
};
