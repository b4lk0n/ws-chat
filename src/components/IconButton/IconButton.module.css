.root {
  --size: 32px;

  display: inline-flex;
  justify-content: center;
  align-items: center;
  width: var(--size);
  height: var(--size);
  user-select: none;
  border: none;
  border-radius: 4px;
  padding: 0;
  outline: none;
  cursor: pointer;
  background: transparent;
  color: var(--primary);
}

.small {
  --size: 18px;
}

.large {
  --size: 40px;
}

.root:disabled {
  opacity: 0.5;
  cursor: default;
}

.root:not(:disabled):not(.small):hover,
.root:not(:disabled):not(.small):focus {
  background-color: var(--outlineLight);
}

.root:not(:disabled):not(.small):active {
  background-color: var(--outline);
}
