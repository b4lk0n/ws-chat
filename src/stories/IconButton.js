import React from 'react';
import { action } from '@storybook/addon-actions';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import { IconButton } from '../components';

export default {
  title: 'IconButton',
  decorators: [withKnobs],
};

export const Default = () => (
  <IconButton
    icon={text('icon', 'add_circle_outline')}
    onClick={action('onClick')}
    size={select('size', {
      'â€”': undefined,
      small: 'small',
      large: 'large',
    })}
  />
);
