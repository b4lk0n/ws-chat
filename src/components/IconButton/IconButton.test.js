import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { IconButton } from './IconButton';

const CustomIcon = () => <span>custom</span>;

describe('<IconButton />', () => {
  it('should render string icon', () => {
    const clickHandler = jest.fn();
    const { container, getByText } = render(<IconButton icon={'account'} onClick={clickHandler} />);

    const icon = getByText('account');
    expect(icon).toBeInTheDocument();
    expect(icon).toHaveClass('material-icons');
    fireEvent.click(container.firstChild);
    expect(clickHandler).toBeCalledTimes(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render custom icon', () => {
    const clickHandler = jest.fn();
    const { container, getByText } = render(<IconButton icon={<CustomIcon />} onClick={clickHandler} />);

    const icon = getByText('custom');
    expect(icon).toBeInTheDocument();
    expect(icon).not.toHaveClass('material-icons');
    fireEvent.click(container.firstChild);
    expect(clickHandler).toBeCalledTimes(1);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should add additional class if provided', () => {
    const { container } = render(<IconButton icon={'account'} className={'bordered'} />);
    expect(container.firstChild).toHaveClass('bordered');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render large icon button', () => {
    const { container, getByText } = render(<IconButton icon={'account'} size={'large'} />);
    expect(container.firstChild).toHaveClass('large');
    expect(getByText('account')).toHaveClass('material-icons');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render small icon button', () => {
    const { container, getByText } = render(<IconButton icon={'account'} size={'small'} />);
    expect(container.firstChild).toHaveClass('small');
    expect(getByText('account')).toHaveClass('material-icons', 'small');
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should not fire events if disabled', () => {
    const clickHandler = jest.fn();
    const { container } = render(<IconButton icon={'account'} onClick={clickHandler} disabled />);
    fireEvent.click(container.firstChild);
    expect(clickHandler).not.toBeCalled();
    expect(container.firstChild).toMatchSnapshot();
  });
});
