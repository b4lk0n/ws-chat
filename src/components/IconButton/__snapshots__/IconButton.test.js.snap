// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IconButton /> should add additional class if provided 1`] = `
<button
  class="root undefined bordered"
>
  <span
    class="root undefined material-icons"
  >
    account
  </span>
</button>
`;

exports[`<IconButton /> should not fire events if disabled 1`] = `
<button
  class="root undefined"
  disabled=""
>
  <span
    class="root undefined material-icons"
  >
    account
  </span>
</button>
`;

exports[`<IconButton /> should render custom icon 1`] = `
<button
  class="root undefined"
>
  <span
    class="root undefined"
  >
    <span>
      custom
    </span>
  </span>
</button>
`;

exports[`<IconButton /> should render large icon button 1`] = `
<button
  class="root large"
>
  <span
    class="root undefined material-icons"
  >
    account
  </span>
</button>
`;

exports[`<IconButton /> should render small icon button 1`] = `
<button
  class="root small"
>
  <span
    class="root small material-icons"
  >
    account
  </span>
</button>
`;

exports[`<IconButton /> should render string icon 1`] = `
<button
  class="root undefined"
>
  <span
    class="root undefined material-icons"
  >
    account
  </span>
</button>
`;
